packages/eslint-plugin/src/rules/strict-boolean-expressions.ts

```diff
@@ -275,6 +367,105 @@ export default createRule<Options, MessageId>({
       if (assertedArgument != null) {
         traverseNode(assertedArgument, true);
       }
+      if (isArrayMethodCallWithPredicate(context, services, node)) {
+        const predicate = node.arguments.at(0);
+        const arrayNode = (node.callee as TSESTree.MemberExpression).object;
+
+        if (predicate) {
+          checkArrayMethodCallPredicate(predicate, arrayNode);
+        }
+      }
+    }
+
+    /**
+     * Dedicated function to check array method predicate calls. Reports predicate
+     * arguments that don't return a boolean value.
+     *
+     * Ignores the `allow*` options and requires a boolean value.
+     */
+    function checkArrayMethodCallPredicate(
+      predicateNode: TSESTree.CallExpressionArgument,
+      arrayNode: TSESTree.Expression,
+    ): void {
+      const isFunctionExpression = isFunctionExpressionNode(predicateNode);
+
+      // custom message for accidental `async` function expressions
+      if (isFunctionExpression && predicateNode.async) {
+        return context.report({
+          node: predicateNode,
+          messageId: 'predicateCannotBeAsync',
+        });
+      }
+
+      const type = checker.getApparentType(
+        services.getTypeAtLocation(predicateNode),
+      );
+
+      const signatures = isFunctionExpression
+        ? type.getCallSignatures()
+        : getMatchingArrayPredicateSignatures(type, arrayNode);
+
+      if (
+        signatures.every(signature =>
+          tsutils
+            .unionTypeParts(signature.getReturnType())
+            .every(isBooleanType),
+        )
+      ) {
+        return;
+      }
+
+      const canFix = isFunctionExpression && !predicateNode.returnType;
+
+      return context.report({
+        node: predicateNode,
+        messageId: 'predicateReturnsNonBoolean',
+        suggest: canFix
```

Comment:
Nice! I've tried to use this here, but I think it's a bit more tricky: The fixer relies on the `canFix` check, which narrows down the type of `predicateNode` to a function expression. Using `getFixOrSuggest` removes that and will require a type assertion, which I'm not sure is isn't worth it.

Wdyt?

Reply:
Oh, that's fun! I'm honestly surprised the narrowing even applies within the closure (it can get reset on mutable things). Anyways, that's reasonable, let's just leave it as is 👍